File: update_books.py
#!/usr/bin/env python3
"""
Book List Update Script (Python Version)

This script helps you quickly commit and push changes to your book list.
Usage: python update_books.py [commit message]

If no commit message is provided, it will use a default one with timestamp.
"""

import subprocess
import sys
import os
from datetime import datetime

# ANSI color codes for pretty output
class Colors:
    GREEN = '\033[32m'
    YELLOW = '\033[33m'
    RED = '\033[31m'
    BLUE = '\033[34m'
    BOLD = '\033[1m'
    RESET = '\033[0m'

def log(message, color=Colors.RESET):
    """Print colored message to console"""
    print(f"{color}{message}{Colors.RESET}")

def run_command(command, description):
    """Run a shell command and handle errors"""
    try:
        log(f"{Colors.BLUE}{description}...{Colors.RESET}")
        result = subprocess.run(
            command, 
            shell=True, 
            capture_output=True, 
            text=True, 
            check=True
        )
        if result.stdout.strip():
            print(result.stdout.strip())
        return True
    except subprocess.CalledProcessError as e:
        log(f"❌ Error: {e.stderr.strip() if e.stderr else str(e)}", Colors.RED)
        return False

def check_git_repo():
    """Check if current directory is a git repository"""
    try:
        subprocess.run(
            "git rev-parse --git-dir", 
            shell=True, 
            capture_output=True, 
            check=True
        )
        return True
    except subprocess.CalledProcessError:
        log("❌ This is not a Git repository. Please run 'git init' first.", Colors.RED)
        return False

def check_books_file():
    """Check if books.md file exists"""
    if not os.path.exists('books.md'):
        log("❌ books.md file not found. Please create it first.", Colors.RED)
        return False
    return True

def get_commit_message():
    """Get commit message from command line args or generate default"""
    if len(sys.argv) > 1:
        return ' '.join(sys.argv[1:])
    
    # Generate default commit message with timestamp
    now = datetime.now()
    date = now.strftime("%b %d, %Y")
    time = now.strftime("%I:%M %p")
    
    return f"Update book list - {date} at {time}"

def has_changes():
    """Check if there are any uncommitted changes"""
    try:
        result = subprocess.run(
            "git status --porcelain", 
            shell=True, 
            capture_output=True, 
            text=True
        )
        return len(result.stdout.strip()) > 0
    except:
        return False

def show_help():
    """Display help information"""
    help_text = f"""
{Colors.BOLD}📚 Book List Update Script (Python){Colors.RESET}

Usage:
  python update_books.py [commit message]

Examples:
  python update_books.py                           # Uses automatic commit message
  python update_books.py "Added 5 new books to wishlist"
  python update_books.py "Finished reading Dune"

This script will:
1. Add all changes to git
2. Create a commit with your message (or auto-generated one)
3. Push the changes to GitHub

Make sure you're in a git repository and have a books.md file.

Options:
  --help, -h    Show this help message
"""
    print(help_text)

def main():
    """Main function"""
    # Check for help flag
    if '--help' in sys.argv or '-h' in sys.argv:
        show_help()
        return
    
    log(f"{Colors.BOLD}📚 Book List Update Script (Python){Colors.RESET}\n")
    
    # Check prerequisites
    if not check_git_repo() or not check_books_file():
        sys.exit(1)
    
    # Check if there are any changes
    if not has_changes():
        log("✨ No changes detected. Your book list is up to date!", Colors.GREEN)
        return
    
    commit_message = get_commit_message()
    log(f"📝 Commit message: \"{commit_message}\"\n")
    
    # Git workflow steps
    steps = [
        ("git add .", "Adding changes to staging area"),
        (f'git commit -m "{commit_message}"', "Creating commit"),
        ("git push", "Pushing to GitHub")
    ]
    
    # Execute each step
    for command, description in steps:
        if not run_command(command, description):
            log("\n❌ Update failed. Please check the error above.", Colors.RED)
            sys.exit(1)
    
    log("\n✅ Book list updated successfully!", Colors.GREEN)
    log("🌐 Your changes should be live on GitHub Pages in a few minutes.", Colors.YELLOW)

if __name__ == "__main__":
    main()

File: update-books.js
#!/usr/bin/env node

/**
 * Book List Update Script
 * 
 * This script helps you quickly commit and push changes to your book list.
 * Run with: node update-books.js [commit message]
 * 
 * If no commit message is provided, it will use a default one with timestamp.
 */

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

// Colors for console output
const colors = {
    green: '\x1b[32m',
    yellow: '\x1b[33m',
    red: '\x1b[31m',
    blue: '\x1b[34m',
    reset: '\x1b[0m',
    bold: '\x1b[1m'
};

function log(message, color = colors.reset) {
    console.log(`${color}${message}${colors.reset}`);
}

function runCommand(command, description) {
    try {
        log(`${colors.blue}${description}...${colors.reset}`);
        const output = execSync(command, { encoding: 'utf8', stdio: 'pipe' });
        if (output.trim()) {
            console.log(output.trim());
        }
        return true;
    } catch (error) {
        log(`❌ Error: ${error.message}`, colors.red);
        return false;
    }
}

function checkGitRepo() {
    try {
        execSync('git rev-parse --git-dir', { stdio: 'pipe' });
        return true;
    } catch {
        log('❌ This is not a Git repository. Please run "git init" first.', colors.red);
        return false;
    }
}

function checkBooksFile() {
    if (!fs.existsSync('books.md')) {
        log('❌ books.md file not found. Please create it first.', colors.red);
        return false;
    }
    return true;
}

function getCommitMessage() {
    const args = process.argv.slice(2);
    
    if (args.length > 0) {
        return args.join(' ');
    }
    
    // Generate default commit message with timestamp
    const now = new Date();
    const date = now.toLocaleDateString('en-US', { 
        year: 'numeric', 
        month: 'short', 
        day: 'numeric' 
    });
    const time = now.toLocaleTimeString('en-US', { 
        hour: '2-digit', 
        minute: '2-digit' 
    });
    
    return `Update book list - ${date} at ${time}`;
}

function hasChanges() {
    try {
        const status = execSync('git status --porcelain', { encoding: 'utf8' });
        return status.trim().length > 0;
    } catch {
        return false;
    }
}

function main() {
    log(`${colors.bold}📚 Book List Update Script${colors.reset}\n`);
    
    // Check prerequisites
    if (!checkGitRepo() || !checkBooksFile()) {
        process.exit(1);
    }
    
    // Check if there are any changes
    if (!hasChanges()) {
        log('✨ No changes detected. Your book list is up to date!', colors.green);
        return;
    }
    
    const commitMessage = getCommitMessage();
    log(`📝 Commit message: "${commitMessage}"\n`);
    
    // Git workflow
    const steps = [
        ['git add .', 'Adding changes to staging area'],
        [`git commit -m "${commitMessage}"`, 'Creating commit'],
        ['git push', 'Pushing to GitHub']
    ];
    
    for (const [command, description] of steps) {
        if (!runCommand(command, description)) {
            log('\n❌ Update failed. Please check the error above.', colors.red);
            process.exit(1);
        }
    }
    
    log('\n✅ Book list updated successfully!', colors.green);
    log('🌐 Your changes should be live on GitHub Pages in a few minutes.', colors.yellow);
}

// Show help if requested
if (process.argv.includes('--help') || process.argv.includes('-h')) {
    console.log(`
📚 Book List Update Script

Usage:
  node update-books.js [commit message]

Examples:
  node update-books.js                    # Uses automatic commit message
  node update-books.js "Added 5 new books to wishlist"
  node update-books.js "Finished reading Dune"

This script will:
1. Add all changes to git
2. Create a commit with your message (or auto-generated one)
3. Push the changes to GitHub

Make sure you're in a git repository and have a books.md file.
`);
    process.exit(0);
}

main();